// ngx-sharebuttons/esm2022/button/share-button.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FontAwesomeModule } from '@fortawesome/angular-fontawesome';
import { ShareModule, SHARE_BUTTONS_CONFIG } from 'ngx-sharebuttons';
import { ShareButton } from './share-button';
import * as i0 from "@angular/core";
class ShareButtonModule {
    static withConfig(config) {
        return {
            ngModule: ShareButtonModule,
            providers: [{ provide: SHARE_BUTTONS_CONFIG, useValue: config }]
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: ShareButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.0.2", ngImport: i0, type: ShareButtonModule, declarations: [ShareButton], imports: [ShareModule,
            FontAwesomeModule,
            CommonModule], exports: [ShareModule,
            FontAwesomeModule,
            ShareButton] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: ShareButtonModule, imports: [ShareModule,
            FontAwesomeModule,
            CommonModule, ShareModule,
            FontAwesomeModule] }); }
}
export { ShareButtonModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: ShareButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ShareButton
                    ],
                    imports: [
                        ShareModule,
                        FontAwesomeModule,
                        CommonModule
                    ],
                    exports: [
                        ShareModule,
                        FontAwesomeModule,
                        ShareButton
                    ]
                }]
        }] });
